src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexFacetedReadOperation.cs

```diff
@@ -0,0 +1,428 @@
+
+public class CoraxIndexFacetedReadOperation : IndexFacetReadOperationBase
+{
+    private readonly IndexFieldsMapping _fieldMappings;
+    private readonly IndexSearcher _indexSearcher;
+    private readonly ByteStringContext _allocator;
+    private readonly Dictionary<string, Slice> _fieldNameCache;
+
+    public CoraxIndexFacetedReadOperation(Index index, Logger logger, Transaction readTransaction, QueryBuilderFactories queryBuilderFactories,
+        IndexFieldsMapping fieldsMapping) : base(index, queryBuilderFactories, logger)
+    {
+        _fieldMappings = fieldsMapping;
+        _allocator = readTransaction.Allocator;
+        _fieldMappings = fieldsMapping;
+        _indexSearcher = new IndexSearcher(readTransaction, _fieldMappings);
+        _fieldNameCache = new();
+    }
+
+    public override List<FacetResult> FacetedQuery(FacetQuery facetQuery, QueryTimingsScope queryTimings, DocumentsOperationContext context,
+        Func<string, SpatialField> getSpatialField, CancellationToken token)
+    {
+        var results = FacetedQueryParser.Parse(context, facetQuery, SearchEngineType.Corax);
+
+        var query = facetQuery.Query;
+        Dictionary<string, Dictionary<string, FacetValues>> facetsByName = null;
+        Dictionary<string, Dictionary<string, FacetValues>> facetsByRange = null;
+
+        var parameters = new CoraxQueryBuilder.Parameters(_indexSearcher, null, null, query, _index, query.QueryParameters, _queryBuilderFactories,
+            _fieldMappings, null, null, -1, null);
+        var baseQuery = CoraxQueryBuilder.BuildQuery(parameters, out var isBinary);
+        var coraxPageSize = CoraxGetPageSize(_indexSearcher, facetQuery.Query.PageSize, query, isBinary);
+        var ids = CoraxIndexReadOperation.QueryPool.Rent(coraxPageSize);
+        using var analyzersScope = new AnalyzersScope(_indexSearcher, _fieldMappings, _index.Definition.HasDynamicFields);
+        int read = 0;
+        while ((read = baseQuery.Fill(ids)) != 0)
+        {
+            for (int docId = 0; docId < read; docId++)
+            {
+                var entryReader = _indexSearcher.GetReaderFor(ids[docId]);
+                foreach (var result in results)
+                {
+                    token.ThrowIfCancellationRequested();
+
+                    using var facetTiming = queryTimings?.For($"{nameof(QueryTimingsScope.Names.AggregateBy)}/{result.Key}");
+
+                    if (result.Value.Ranges == null || result.Value.Ranges.Count == 0)
+                    {
+                        facetsByName ??= new Dictionary<string, Dictionary<string, FacetValues>>();
+
+                        HandleFacetsPerDocument(ref entryReader, result, facetsByName, facetQuery.Legacy, facetTiming, analyzersScope, token);
+                        continue;
+                    }
+
+                    // Cache facetByRange because we will fulfill data in batches instead of whole collection
+                    facetsByRange ??= new();
```

Comment:
You can do this in initialization, no? Saves a branch per result

Reply:
yes, done